---
title: "AustinM_EG_AsstNo2_code.Rmd"
author: "Maia Austin"
date: "10/25/2021"
output: pdf_document
---
```{r}
library(DESeq2) 
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(wesanderson)
library(vsn)
library(hexbin)
library(eulerr)
library(pheatmap)
setwd("~/Documents/GitHub/EcologicalGenomics/Assignment2")
```
# F1 generation analysis
## Import
```{r}
setwd("~/Documents/GitHub/EcologicalGenomics/Assignment2/tonsaF1")
# import counts matrix
countsTable <- read.table("DE_counts_F1.txt", header=T, row.names = 1)
head(countsTable)
dim(countsTable) # 24362  x  16

countsTableRound <- round(countsTable) #removing decimals
head(countsTableRound)

# import the sample description table
conds <- read.delim("RT_tonsa_F1_samples.txt", header=T, stringsAsFactors = T, row.names = 1)
head(conds)

# import the sample description table
conds <- read.delim("RT_tonsa_F1_samples.txt", header=T, stringsAsFactors = T, row.names = 1)
head(conds)
```

## Preliminary Data Exploration
```{r}
#check how many reads we have from each sample
colSums(countsTableRound)
mean(colSums(countsTableRound))
# just over 18 million reads per sample

#The average number counts per gene
rowSums(countsTableRound)
mean(rowSums(countsTableRound)) #11930.81
median(rowSums(countsTableRound)) #2226
#median much lower than mean; expected with distribution with long tail

```

## DESeq
```{r}
### create a DESeq object and define experimental design here with the tilda

dds <- DESeqDataSetFromMatrix(countData=countsTableRound, 
                              colData=conds, 
                              design=~line+environment+line:environment) #[cond]:[cond] = interaction

dim(dds)

### Filter out genes with too few reads - keep reads with average > 10 reads per sample

dds <- dds[rowSums(counts(dds)) > 160] #10 * 16 samples
dim(dds)
     
### Run DESeq Model to test for differential gene expression
dds <- DESeq(dds)

### List the results you've generated
resultsNames(dds)
# [1] "Intercept"                 
# [2] "line_combined_vs_ambient"  
# [3] "environment_HH_vs_AA"      
# [4] "linecombined.environmentHH"
```

## Results from specific contrasts
### Environment Effect
```{r}

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, 
                              colData = conds, 
                              design = ~ line + environment)
dds <- DESeq(dds, test="LRT", reduced=~line)
resultsNames(dds)
# [1] "Intercept"               
# [2] "line_combined_vs_ambient"
# [3] "environment_HH_vs_AA"   

resEnv <- results(dds, alpha = 0.05)
resEnv <- resEnv[order(resEnv$padj),] #ordering rows by adjusted p-value
head(resEnv)

summary(resEnv)
# out of 24362 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 213, 0.87%
# LFC < 0 (down)     : 235, 0.96%
# outliers [1]       : 41, 0.17%
# low counts [2]     : 473, 1.9%
# (mean count < 18)

resEnv <- resEnv[!is.na(resEnv$padj),] #filters data table to exclude NA results; lost about 500 genes

degsEnv <- row.names(resEnv[resEnv$padj < 0.05,]) #differentially expressed genes from the environment
length(degsEnv) 
#448 genes differentially expressed
```

### Line Effect
```{r}
dds <- DESeqDataSetFromMatrix(countData = countsTableRound, 
                              colData = conds, 
                              design = ~ environment + line)

dds <- DESeq(dds, test="LRT", reduced=~environment)
resultsNames(dds)
# [1] "Intercept"               
# [2] "environment_HH_vs_AA"    
# [3] "line_combined_vs_ambient"

resLine <- results(dds, alpha = 0.05)
resLine <- resLine[order(resLine$padj),]
head(resLine)
summary(resLine)
# out of 24362 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 72, 0.3%
# LFC < 0 (down)     : 154, 0.63%
# outliers [1]       : 41, 0.17%
# low counts [2]     : 2362, 9.7%
# (mean count < 25)
resLine <- resLine[!is.na(resLine$padj),]

degsline <- row.names(resLine[resLine$padj < 0.05,])
length(degsline)
#226 differentially expressed gene
```

### Interaction Effect
```{r}
dds <- DESeqDataSetFromMatrix(countData = countsTableRound, 
                              colData = conds, 
                              design = ~ environment + line + environment:line)

dds <- DESeq(dds, test="LRT", reduced=~environment + line)
resultsNames(dds)
# [1] "Intercept"                 
# [2] "environment_HH_vs_AA"      
# [3] "line_combined_vs_ambient"  
# [4] "environmentHH.linecombined"
resInt <- results(dds, alpha = 0.05)
resInt <- resInt[order(resInt$padj),]
head(resInt)

summary(resInt)
# out of 24362 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 2802, 12%
# LFC < 0 (down)     : 1052, 4.3%
# outliers [1]       : 9, 0.037%
# low counts [2]     : 945, 3.9%
# (mean count < 20)

resInt <- resInt[!is.na(resInt$padj),]

degsInt <- row.names(resInt[resInt$padj < 0.05,])
length(degsInt)
#3854 differentially expressed genes
```

## Data Visualization

### Venn Diagram
```{r}
# Total
length(degsEnv)  # 448
length(degsline)  # 226
length(degsInt)  # 3854

# Intersections
length(intersect(degsEnv,degsline))  # 37
length(intersect(degsEnv,degsInt))  # 44
length(intersect(degsInt,degsline))  # 34

intEL <- intersect(degsEnv,degsline)
length(intersect(degsInt,intEL)) # 7

# Number unique
448-44-37-7 # 360 
226-37-34-7 # 148
3854-44-34-7 # 3769

# unique, present in 2, present in 3
fit1 <- euler(c("Env" = 360, "Line" = 148, "Interaction" = 3769, "Env&Line" = 37, "Env&Interaction" = 44, "Line&Interaction" = 34, "Env&Line&Interaction" = 7))

plot(fit1,  lty = 1:3, quantities = TRUE)

plot(fit1, quantities = TRUE, fill = "transparent",
     lty = 1:3,
     labels = list(font = 4))
```

### Interaction Heatmap
```{r}
vsd <- vst(dds, blind=F)
topgenes <- head(rownames(resInt),50)
mat <- assay(vsd)[topgenes,]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds)[,c("line","environment")])
pheatmap(mat, annotation_col=df, fontsize_row = 6)
#heatmap colors - up and downregulation
```

# F3 Generation Analysis
## Import
```{r}
setwd("~/Documents/GitHub/EcologicalGenomics/Assignment2/tonsaF3")
# import counts matrix
countsTable <- read.table("DE_counts_F3.txt", header=T, row.names = 1)
head(countsTable)
dim(countsTable) # 25279  x  16

countsTableRound <- round(countsTable) #removing decimals
head(countsTableRound)

# import the sample description table
conds <- read.delim("RT_tonsa_F3_samples.txt", header=T, stringsAsFactors = T, row.names = 1)
head(conds)

```

## Preliminary Data Exploration
```{r}
#check how many reads we have from each sample
colSums(countsTableRound)
mean(colSums(countsTableRound))
# just over 18 million reads per sample

#The average number counts per gene
rowSums(countsTableRound)
mean(rowSums(countsTableRound)) #11220.54
median(rowSums(countsTableRound)) #2144
#median much lower than mean; expected with distribution with long tail

```

## DESeq
```{r}
### create a DESeq object and define experimental design here with the tilda

dds <- DESeqDataSetFromMatrix(countData=countsTableRound, 
                              colData=conds, 
                              design=~line+environment+line:environment) #[cond]:[cond] = interaction

dim(dds)

### Filter out genes with too few reads - keep reads with average > 10 reads per sample

dds <- dds[rowSums(counts(dds)) > 160] #10 * 16 samples
dim(dds)
     
### Run DESeq Model to test for differential gene expression
dds <- DESeq(dds)

### List the results you've generated
resultsNames(dds)
# [1] "Intercept"                 
# [2] "line_combined_vs_ambient"  
# [3] "environment_HH_vs_AA"      
# [4] "linecombined.environmentHH"
```

## Results from specific contrasts
### Environment Effect
```{r}

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, 
                              colData = conds, 
                              design = ~ line + environment)
dds <- DESeq(dds, test="LRT", reduced=~line)
resultsNames(dds)
# [1] "Intercept"               
# [2] "line_combined_vs_ambient"
# [3] "environment_HH_vs_AA"   

resEnv <- results(dds, alpha = 0.05)
resEnv <- resEnv[order(resEnv$padj),] #ordering rows by adjusted p-value
head(resEnv)

summary(resEnv)
# out of 25279 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 513, 2%
# LFC < 0 (down)     : 315, 1.2%
# outliers [1]       : 16, 0.063%
# low counts [2]     : 491, 1.9%
# (mean count < 19)

resEnv <- resEnv[!is.na(resEnv$padj),] #filters data table to exclude NA results; lost about 500 genes

degsEnv <- row.names(resEnv[resEnv$padj < 0.05,]) #differentially expressed genes from the environment
length(degsEnv) 
#828 genes differentially expressed
```

### Line Effect
```{r}
dds <- DESeqDataSetFromMatrix(countData = countsTableRound, 
                              colData = conds, 
                              design = ~ environment + line)

dds <- DESeq(dds, test="LRT", reduced=~environment)
resultsNames(dds)
# [1] "Intercept"               
# [2] "environment_HH_vs_AA"    
# [3] "line_combined_vs_ambient"

resLine <- results(dds, alpha = 0.05)
resLine <- resLine[order(resLine$padj),]
head(resLine)
summary(resLine)
# out of 25279 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 808, 3.2%
# LFC < 0 (down)     : 837, 3.3%
# outliers [1]       : 16, 0.063%
# low counts [2]     : 981, 3.9%
# (mean count < 21)
resLine <- resLine[!is.na(resLine$padj),]

degsline <- row.names(resLine[resLine$padj < 0.05,])
length(degsline)
#1645 differentially expressed gene
```

### Interaction Effect
```{r}
dds <- DESeqDataSetFromMatrix(countData = countsTableRound, 
                              colData = conds, 
                              design = ~ environment + line + environment:line)

dds <- DESeq(dds, test="LRT", reduced=~environment + line)
resultsNames(dds)
# [1] "Intercept"                 
# [2] "environment_HH_vs_AA"      
# [3] "line_combined_vs_ambient"  
# [4] "environmentHH.linecombined"
resInt <- results(dds, alpha = 0.05)
resInt <- resInt[order(resInt$padj),]
head(resInt)

summary(resInt)
# out of 25279 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 235, 0.93%
# LFC < 0 (down)     : 48, 0.19%
# outliers [1]       : 5, 0.02%
# low counts [2]     : 2941, 12%
# (mean count < 28)

resInt <- resInt[!is.na(resInt$padj),]

degsInt <- row.names(resInt[resInt$padj < 0.05,])
length(degsInt)
#283 differentially expressed genes
```

## Data Visualization

### Venn Diagram
```{r}
length(degsEnv)  # 828
length(degsline)  # 1645
length(degsInt)  # 283

# Intersections
length(intersect(degsEnv,degsline))  # 141
length(intersect(degsEnv,degsInt))  # 14
length(intersect(degsInt,degsline))  # 32

intEL <- intersect(degsEnv,degsline)
length(intersect(degsInt,intEL)) # 7

# Number unique
828-14-141-7 # 666
1645-141-32-7 # 1465
283-14-32-7 # 230

# unique, present in 2, present in 3
fit1 <- euler(c("Env" = 666, "Line" = 1465, "Interaction" = 230, "Env&Line" = 141, "Env&Interaction" = 14, "Line&Interaction" = 32, "Env&Line&Interaction" = 7))

plot(fit1,  lty = 1:3, quantities = TRUE)

plot(fit1, quantities = TRUE, fill = "transparent",
     lty = 1:3,
     labels = list(font = 4))
```

### Interaction Heatmap
```{r}
vsd <- vst(dds, blind=F)
topgenes <- head(rownames(resInt),50)
mat <- assay(vsd)[topgenes,]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds)[,c("line","environment")])
pheatmap(mat, annotation_col=df, fontsize_row = 6)
#heatmap colors - up and downregulation
```