---
title: "Ecological Genomics Notes"
author: "Maia Austin"
date: "9/1/2021"
output: html_document
---
# MICROBIOME

## Tutorial 1 
GitHub Setup
Access server: ssh netID@pbio381.uvm.edu

## Tutorial 2
wd: pwd
change directories: cd /data/ (root)
change directories (not root): cd project_data/
list files and directories in wd: ll
go back to last place: cd ..
autofill: type in part of title (pr...) then hit tab
arrow up/down: navigate between past commands
16s data: /data/project_data/16s
get into gz files: zcat SS15_08_R1.fq.gz | head -n4
  4 lines: 
    1. info about read (begins with '@')
    2. DNA sequence, 
    3. more metadata (begins with '+')
    4. quality info (same # characters as line 2)

access file: less [file_name] OR vim [file_name]
quit: :q
stop text wrapping: :set nowrap
go back to home directory: cd~
make directory: mkdir [name]
Activate qiime: 

ssh maustin7@pbio381.uvm.edu
conda activate qiime2-2021.8



qiime tools import \
  --type 'SampleData[PairedEndSequencesWithQuality]' \
  --input-path /data/project_data/16S/pyc_subset_manifest \
  --input-format PairedEndFastqManifestPhred33V2 \
  --output-path demux-paired-end_full.qza


## Tutorial 3

#### View qiime file:

qiime demux summarize \
  --i-data demux-paired-end.qza \      
  --o-visualization demux-pyc-sub.qzv


#### File transfer:

scp maustin7@pbio381.uvm.edu:~/myresults/*.qzv .

Enter screen: screen [enter]
  deattach before closing: [ctrl][a][d]
  reattach: screen -r


#### Denoise:

qiime dada2 denoise-paired \
  --i-demultiplexed-seqs demux-paired-end_full.qza \
  --p-n-threads 1 \
  --p-trim-left-f 16 \
  --p-trim-left-r 0 \
  --p-trunc-len-f 278 \
  --p-trunc-len-r 232 \
  --o-table table.qza \
  --o-representative-sequences rep-seqs.qza \
  --o-denoising-stats denoising-stats.qza


Forward - 16-278
Reverse - 0-232

#### View denoising info:

qiime feature-table summarize \
  --i-table table.qza \
  --o-visualization table.qzv \
  --m-sample-metadata-file /data/project_data/16S/pyc_manifest

qiime feature-table tabulate-seqs \
  --i-data rep-seqs.qza \
  --o-visualization rep-seqs.qzv

qiime metadata tabulate \
  --m-input-file denoising-stats.qza \
  --o-visualization denoising-stats.qzv


## Tutorial 4
Decide sampling depth by looking at table.qzv

### What questions can we ask/address with these data?
1. lower alpha and higher beta in sick
2. less diversity in sick
3. more maladaptive in sick
4. does site or health status impact diversity/composition more
5. shift from healthy to sick
6. healthy at impacted vs healthy at naive
7. site specific healthy vs. sick

#### Build phylogenetic tree:

qiime phylogeny align-to-tree-mafft-fasttree \
  --i-sequences rep-seqs.qza \
  --o-alignment aligned-rep-seqs.qza \
  --o-masked-alignment masked-aligned-rep-seqs.qza \
  --o-tree unrooted-tree.qza \
  --o-rooted-tree rooted-tree.qza

#### Diversity matrices:

qiime diversity core-metrics-phylogenetic \
  --i-phylogeny rooted-tree.qza \
  --i-table table.qza \
  --p-sampling-depth 18000 \
  --m-metadata-file pyc_manifest \
  --output-dir core-metrics-results

Sampling depth: 18000

qiime diversity alpha-group-significance \
  --i-alpha-diversity core-metrics-results/faith_pd_vector.qza \
  --m-metadata-file pyc_manifest \
  --o-visualization core-metrics-results/faith-pd-group-significance.qzv
  
qiime diversity alpha-group-significance \
  --i-alpha-diversity core-metrics-results/evenness_vector.qza \
  --m-metadata-file pyc_manifest \
  --o-visualization core-metrics-results/evenness-group-significance.qzv

qiime diversity beta-group-significance \
  --i-distance-matrix core-metrics-results/weighted_unifrac_distance_matrix.qza \
  --m-metadata-file pyc_manifest \
  --m-metadata-column site-animal-health \
  --o-visualization core-metrics-results/weighted-unifrac-site-animal-health-significance.qzv \
  --p-pairwise

qiime diversity alpha-rarefaction \
  --i-table table.qza \
  --i-phylogeny rooted-tree.qza \
  --p-max-depth 100000 \
  --m-metadata-file pyc_manifest \
  --o-visualization alpha-rarefaction.qzv

#### Taxonomic ID
https://pespenilab.github.io/Ecological-Genomics/2021_09_27_microbiome_supp_taxaID.html

qiime feature-classifier classify-sklearn \
  --i-classifier /data/project_data/16S/training-feature-classifiers/classifier.qza \
  --i-reads /users/m/a/maustin7/myresults/rep-seqs.qza \
  --o-classification taxonomy.qza
  
qiime taxa barplot \
  --i-table table.qza \
  --i-taxonomy /users/m/a/maustin7/training-feature-classifiers/taxonomy.qza \
  --m-metadata-file pyc_manifest \
  --o-visualization taxa-bar-plots.qzv
  
Note: look to tutorial for more stats on abundance

# RNASeq
## Tutorial 1
Identify SNPs from RNASeq data
Looking at 4 conditions of Acartia Hudsonica:
1. ocean acidification (1000 micro-atm pCO2, ambient temperature, 13 degrees C)
2. ocean warming (15 degrees C, 400 micro-atm pCO2)
3. combined ocean acidification and warming (1000 micro-atm pCO2, 15 degrees C)
4. ambient (13 degrees C, 400 micro-atm pCO2)

Questions:
1. acidification vs warming effect, and combined - additive? synergistic?
2. generational changes, gene expression shifts - what happens between each generation?
3. time course of gene expression or alleles
4. Using busco (sp?) to tell how complete transcriptome assembly is

fastq = .fq.gz

Processing pipeline:
1. fastq -> FastQC -> html -> local machine (quality report)
2. fastq -> Trimmomatic -> clean.fq.gz files (trimming based on quality report)
3. clean.fq.gz -> FastQC (make sure trimming worked)
4. clean.fq.gz -> Trinity -> ref.fa (reference transcriptome)
5. ref.fa -> quality checks, etc. before it's the final one
6. ref.fa + clean.fq.gz -> Salmon -> counts.txt (data frame of samples and genes)
  a. counts.txt - Number of reads in sample that match given transcriptome
7. counts.txt -> DESeq2 (R program, test differences in gene expression)
8. ref.fa + clean.fq.gz -> bwamem -> .sam (sequence alignment files) -> .bam
  a. looking at SNPs at given nucleotide position, get freqeuncy of each allele within sample
9. .bam -> program like VarScan (scan for variants) -> .vcf (entry for every polymorphism)
10. .vcf -> CMH test statistic (looks for consistent differences in allele frequency)

doing Warming F2 (HA_F2)

Wildcard - *, grabs anything with that
  can check with echo

fastqc HA_F2* --outdir=/data/project_data/RNAseq/fastqc

## Tutorial 2

### Make a script:
vim [title].sh
'i' for insert/to start type
#!/bin/bash

Leave and write file:
[esc]
:wq!

black=non-executable, green=executable
make it executable: chmod u+x test.sh

Run script: ./test.sh

### Check number of html files
ll *.html | wc -l

## Workflow:
raw fq.qz -> FastQC -> html -> MultiQC -> .html
raw fq.qz -> Trimmomatic -> (clean) qc.fq.qz -> FastQC -> MultiQC

Clean data seems to still have hexamer primers at start - messing with per base sequence data in MultiQC
  Melissa fixed! did HEADCROP on 12 bases (should have removed)
  
Cleaned FastQ -> Trinity/Transcriptome assembly 
Transcriptome assembly + cleaned fastq -> Map/align reads to transcriptome

Mapping -
  -Salmon -> alignment free mapping (uses kmers/unique sets of nucleotides to find probability wise where transcriptome will fall)
  -Alignment (like bwa and bowtie2) -> .sam files; actually aligning and finding matches
  
Can also map to well-annotated genome or closely related species, instead of transcriptome

We're going to be looking at counts.txt generation 21 of transplant experiment while troubleshooting is happening
AA - ambient condition, HH - high temp high c02

See deseq2 R script for instructions!


Model - simplified representation of reality

counts ~ line, environment, interaction(line:environment)
Negative binomial GLM (generalized linear model)

LRT - Full or reduced model

# PopGen
## Tutorial 1 - Background
Module 1 - microbiome
  -assessing species/taxa level diversity
  -how environment shapes community diversity
  -AmpliconSeq approach
Module 2 - Transcriptomics
  -RNASeq
  -assay across transcriptome to get levels of gene expression
  -responses to experimental evolution and environmental treatment
Module 3 - population and landscape genomics
  -assessing levels of genomic diversity across genome (DNA, not RNA)
  -whole genome sequencing
  -library prep across whole genome!

What defines a population?
  high degree of mating (random) within group -> recombination
  topographical barriers
  pre- and post-zygotic barriers? (flowering time)
  genetic drift (small effective population size (Ne), more likely to undergo genetic drift)
  local adaptation -> hybrids selected against

Genetic landscape (across chromosome) is very heterogeneous
  Tend to see selection at areas with extremes (high expression or low expression)
  Hitchhiking - nearby/linked variants to selected variant change with variant
  
Fst - how much two populations differ
  s = subpopulation, t = total, F = inbreeding
  spatial structure causes type of inbreeding
  H(T) - H(S) / H(T) (heterozygosity in total pop. vs. subpop)
  0 > Fst > 1
  
Islands of divergence between populations within species/taxa, often said to be contributing to reproductive isolation

Gets complicated very fast - rare to get to absolute certainty

look for papers that talk about some of these dynamics (gene flow/recombination in spatial contexts, local adaptation)  
  
Looking at poplar tree hybrid zone and admixture
Ecologically and genetically diverse
Compact genome (about 1/6 size of human genome)
Diploid

What questions can we ask?
1. is there a genomic correlation with the climate data? Is there a clear hybrid zone? 
2. Differnetiation along transects 
3. phylogenetic trees?
4. plasticity amongst transplant sites
5. succession/forest fire effects

Processing Pipeline (pre-class)
fastq files - [Trimmomatic (remove adapters, trim bp < Q30)] -> cleaned fastq - [bwa (take reads and align/map to reference)] -> bam (binary alignment files) - [GATK] -> VCF (variant call format) - [VCFTools (filtering by minor allele frequency)] -> filtered VCF - [BEAGLE (imputation/filling in missing values)] -> Final VCF!

VCF - just variant sites; tells you specific BP, if homo/heterozygote for reference allele and for variant allele

Minor allele frequency - whichever allele at a given site is the rarer of the two; map between 0 and 0.5, will have more around 0
  -sequencing errors will occur around 0; need to filter!
  -our cutoff - 0.002

10% missingness - given SNP position needed to be present in at least 90% of the samples

## Tutorial 2
view zipped file head - zcat poplar_hybrids.vcf.gz | head -n 11
  | pipe - merge 2 commands
  
vcftools --[filetype] [filename]
  vcftools --gzvcf poplar_hybrids.vcf.gz
  NOT changing file, just looking
We have 576 samples - ~ 5 million SNPs
Most people will go between 0.08 and 0.05 - very little difference
Lowest minor allele frequency will be  (1/2*sample number)

grep - regular expression command; search for pattern
grep "Alaska" Combined_Transect_Sampling_Data_2020.txt | cutoff -f1
  Print all sample IDs from Alaska transect

vcftools --gzvcf poplar_hybrids.vcf.gz --maf 0.05 --recode --out poplar_hybrids_maf05

k = 3
see bashscript admixturetest.sh - missing a few steps for directory manip here

Thin out linkage disequilibrium:
plink2 --vcf $VCF \
--threads 3 \
--double-id \ 
--allow-extra-chr \
--set-missing-var-ids @:# \
--indep-pairwise 50 10 0.1 \ #using window of 50,000 bp, thin out any with LD score greater than 0.1
--out poplar_hybrids_noLD

convert from VCF to admixture:
plink2 --vcf $VCF \
--threads 3 \
--allow-extra-chr \
--make-bed \
--out Admixture/$FILE 

subset for no LD SNPs:
plink2 --bfile Admixture/$FILE \ 
--threads 3 \
--allow-extra-chr \
--set-missing-var-ids @:# \
--extract poplar_hybrids_noLD.prune.in \
--make-bed \
--out Admixture/$FILE.LDpruned 

Run Admixture:
K=<your value>
admixture -j3 --cv $FILE2.bed $K >log${K}.out

Can do loop of values:
for K in {2..5}
do
 admixture -j3 --cv $FILE2.bed $K >log${K}.out
done

Take .Q file
Input into R script:
library(maps)  # If you don't have this library, you'll need to run install.packages("maps") first
library(plotrix) # If you don't have this library, you'll need to run install.packages("plotrix") first

setwd()  #set the path to where your downloaded files are on your laptop

meta <- read.table("Combined_Transect_Sampling_Data_2020.txt", sep="\t",header=T)  

Qscores <- read.table("yourfilename.Q", sep=" ",header=F)
names(Qscores) <- c("K1","K2", ...)  # Customize to your level of K!

tiff("Admix.tiff", width = 10, height = 7, units = "in",
     res = 300)
map("world",xlim=c(-160,-100),ylim=c(40,70),fill=T, col="lightgrey")

title("Admixture K=X") # Change for your value of K
map.axes(cex.axis=1.5)
for(i in 1:nrow(meta)){
  floating.pie(meta$Longitude[i], meta$Latitude[i], 
               c(Qscores$K1[i],Qscores$K2[i],...),
               col=c("yellow","green",...),
               radius=0.5)
}

# Customize the above to your level of K wherever you see the ellipses (...)!

dev.off()

CROSS VALIDATION ERROR
grep "CV" log3.out

K   CV
2   0.5146
3   0.50303
4   0.49744
5   0.49513
6   0.49514
7   0.49567
8   0.49540

The higher the admixture (diversity of ancestry scores), the more genome-wide heterozygosity we expect to see

Going to use K=5 - lowest cv error
see k5het.R script